File: edp_go/philosopher.go
package main

import (
	"context"
	"fmt"
	"sync"
	"time"
)

type Philosopher struct {
	name                Name
	money               float64
	seat                *Seat
	order               *Order
	foodServedChan      chan bool
	shouldReceiveCoupon bool
	ctx                 context.Context
	cancel              context.CancelFunc
	wg                  sync.WaitGroup
}

var philosophers []*Philosopher
var philosophersMutex sync.RWMutex

func init() {
	philosophers = make([]*Philosopher, len(PHILOSOPHER_NAMES))
	for i, name := range PHILOSOPHER_NAMES {
		ctx, cancel := context.WithCancel(context.Background())
		philosophers[i] = &Philosopher{
			name:           name,
			money:          STARTING_MONEY,
			foodServedChan: make(chan bool, 1),
			ctx:            ctx,
			cancel:         cancel,
		}
	}
}

func ListPhilosophers() []*Philosopher {
	philosophersMutex.RLock()
	defer philosophersMutex.RUnlock()
	result := make([]*Philosopher, len(philosophers))
	copy(result, philosophers)
	return result
}

func GetHeldChopsticks() map[int]bool {
	heldChopsticksMutex.RLock()
	defer heldChopsticksMutex.RUnlock()
	result := make(map[int]bool)
	for k, v := range heldChopsticks {
		result[k] = v
	}
	return result
}

func (p *Philosopher) Name() Name            { return p.name }
func (p *Philosopher) Order() *Order         { return p.order }
func (p *Philosopher) SetOrder(order *Order) { p.order = order }
func (p *Philosopher) ClearOrder()           { p.order = nil }

func (p *Philosopher) think() { time.Sleep(THINKING_TIME) }

func (p *Philosopher) eat() {
	seatNumber := p.seat.Number()
	leftChopstick := seatNumber
	rightChopstick := (seatNumber + 1) % len(philosophers)

	var firstChopstick, secondChopstick int
	if seatNumber == len(philosophers)-1 {
		firstChopstick, secondChopstick = rightChopstick, leftChopstick
	} else {
		firstChopstick, secondChopstick = leftChopstick, rightChopstick
	}

	CHOPSTICKS[firstChopstick].Lock()
	heldChopsticksMutex.Lock()
	heldChopsticks[firstChopstick] = true
	heldChopsticksMutex.Unlock()

	CHOPSTICKS[secondChopstick].Lock()
	heldChopsticksMutex.Lock()
	heldChopsticks[secondChopstick] = true
	heldChopsticksMutex.Unlock()

	time.Sleep(EATING_TIME)

	heldChopsticksMutex.Lock()
	delete(heldChopsticks, secondChopstick)
	heldChopsticksMutex.Unlock()
	CHOPSTICKS[secondChopstick].Unlock()

	heldChopsticksMutex.Lock()
	delete(heldChopsticks, firstChopstick)
	heldChopsticksMutex.Unlock()
	CHOPSTICKS[firstChopstick].Unlock()
}

func (p *Philosopher) pay() {
	mealCost := p.order.Cost()
	p.money -= mealCost
	if p.money < 0 {
		fmt.Printf("Philosopher %s cannot afford the meal ($%.2f) and is leaving for good. Balance: $%.2f\n",
			p.name, mealCost, p.money)
		p.VacateSeat()
		p.money = 0
	} else {
		fmt.Printf("Philosopher %s has paid $%.2f and left the restaurant.\n", p.name, mealCost)
		p.VacateSeat()
	}
}

func (p *Philosopher) waitForWaiter() bool {
	fmt.Printf("Philosopher %s is waiting for a waiter.\n", p.name)
	waiterManager.AddPhilosopherToQueue(p)

	startTime := time.Now()
	for time.Since(startTime) <= TIMEOUT {
		if p.order != nil {
			fmt.Printf("Philosopher %s got an order from waiter.\n", p.name)
			return true
		}
		time.Sleep(WAITING_TIME)
	}

	fmt.Printf("Philosopher %s gave up waiting for a waiter.\n", p.name)
	return false
}

func (p *Philosopher) NotifyFoodServed() {
	select {
	case p.foodServedChan <- true:
	default:
	}
}

func (p *Philosopher) SetShouldReceiveCoupon(shouldReceive bool) {
	p.shouldReceiveCoupon = shouldReceive
}

func (p *Philosopher) giveCoupon(amount float64) {
	p.money += amount
	fmt.Printf("Philosopher %s received a $%.2f coupon. New balance: $%.2f\n", p.name, amount, p.money)
}

func (p *Philosopher) Run() {
	defer p.wg.Done()
	defer p.cancel()
	for p.money > 0 {
		p.attemptToDine()
		p.think()
	}
	p.VacateSeat()
	fmt.Printf("Philosopher %s has left the restaurant for good.\n", p.name)
}

func (p *Philosopher) attemptToDine() {
	p.seat = seatManager.AvailableSeat()
	if p.seat != nil && p.seat.AttemptToOccupy() {
		fmt.Printf("Philosopher %s is being seated in chair %d.\n", p.name, p.seat.Number())
		p.think()
		fmt.Printf("Philosopher %s is about to call for a waiter.\n", p.name)
		hasWaiter := p.waitForWaiter()
		if !hasWaiter {
			fmt.Printf("Philosopher %s has left the restaurant without being served.\n", p.name)
			p.seat.Vacate()
			return
		}
		fmt.Printf("Philosopher %s got waiter, waiting for food.\n", p.name)
		p.shouldReceiveCoupon = false
		select {
		case <-p.foodServedChan:
			if p.shouldReceiveCoupon {
				p.giveCoupon(COUPON_VALUE)
				return
			}
		case <-p.ctx.Done():
			return
		}
		fmt.Printf("Philosopher %s got food, about to eat.\n", p.name)
		p.eat()
		if p.order != nil {
			p.pay()
			p.ClearOrder()
		}
	} else {
		fmt.Printf("Philosopher %s could not get a seat.\n", p.name)
	}
}

func (p *Philosopher) VacateSeat() {
	if p.seat != nil {
		p.seat.Vacate()
		p.seat = nil
	}
}

File: edp_java/src/main/Philosopher.java
package main;

import java.util.Optional;
import java.util.concurrent.CountDownLatch;

public class Philosopher extends Thread {

    // Constants (Time in milliseconds)
    private static final long THINKING_TIME = 1000;
    private static final long EATING_TIME = 2500;
    private static final long WAITING_TIME = 100;
    private static final long TIMEOUT = 2500;
    private static final float STARTING_MONEY = 200.0f;
    private static final float COUPON_VALUE = 5.0f;

    private static final Philosopher[] PHILOSOPHERS = java.util.Arrays.stream(Name.PHILOSOPHER_NAMES)
            .map(n -> new Philosopher(n.toString())).toArray(Philosopher[]::new);

    // Track which chopsticks are currently being held
    private static final java.util.Set<Integer> heldChopsticks = java.util.concurrent.ConcurrentHashMap.newKeySet();

    public static Philosopher[] list() {
        return PHILOSOPHERS;
    }

    public static java.util.Set<Integer> getHeldChopsticks() {
        return new java.util.HashSet<>(heldChopsticks);
    }

    private final Name name;
    private float money = STARTING_MONEY;
    private Optional<Seat> seat = Optional.empty();
    private Optional<Order> order = Optional.empty();
    private CountDownLatch foodServedLatch;
    private volatile boolean shouldReceiveCoupon = false;

    public Name name() {
        return name;
    }

    public Philosopher(String name) {
        this.name = new Name(name);
        Utility.validateTime(TIMEOUT);
    }

    public Optional<Order> order() {
        return order;
    }

    public void setOrder(Order order) {
        this.order = Optional.of(order);
    }

    public void clearOrder() {
        this.order = Optional.empty();
    }

    private void think() {
        try {
            Thread.sleep(THINKING_TIME);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    private void eat() {
        int seatNumber = seat.get().number();
        int leftChopstick = seatNumber;
        int rightChopstick = (seatNumber + 1) % PHILOSOPHERS.length;

        int firstChopstick;
        int secondChopstick;
        // Last philosopher picks up right chopstick first, then left
        if (seatNumber == PHILOSOPHERS.length - 1) {
            firstChopstick = rightChopstick;
            secondChopstick = leftChopstick;
        } else {
            firstChopstick = leftChopstick;
            secondChopstick = rightChopstick;
        }

        synchronized (EnhancedDiningPhilosophers.CHOPSTICKS[firstChopstick]) {
            heldChopsticks.add(firstChopstick);
            synchronized (EnhancedDiningPhilosophers.CHOPSTICKS[secondChopstick]) {
                heldChopsticks.add(secondChopstick);
                Utility.waitFor(EATING_TIME);
                heldChopsticks.remove(secondChopstick);
            }
            heldChopsticks.remove(firstChopstick);
        }
    }

    private void pay() {
        float mealCost = order.get().cost();
        money -= mealCost;
        if (money < 0) {
            System.out.println(String.format(
                    "Philosopher %s cannot afford the meal ($%.2f) and is leaving for good. Balance: $%.2f",
                    name, mealCost, money));
            money = 0;
        } else {
            System.out.println(
                    String.format("Philosopher %s has paid $%.2f and left the restaurant.", name, mealCost));
        }
        vacateSeat();
    }

    private boolean waitForWaiter() {
        System.out.println(String.format("Philosopher %s is waiting for a waiter.", name));

        // Add to the blocking queue
        Waiter.addPhilosopherToQueue(this);

        // Wait for a waiter to take the order within TIMEOUT
        long startTime = System.currentTimeMillis();
        while (System.currentTimeMillis() - startTime <= TIMEOUT) {
            // Check if we have an order (meaning a waiter took our order)
            if (order.isPresent()) {
                System.out.println(String.format("Philosopher %s got an order from waiter.", name));
                return true;
            }
            Utility.waitFor(WAITING_TIME);
        }

        // Timeout reached - remove from queue if still there
        Waiter.removePhilosopherFromQueue(this);
        System.out.println(String.format("Philosopher %s gave up waiting for a waiter.", name));
        return false;
    }

    public void notifyFoodServed() {
        if (foodServedLatch != null)
            foodServedLatch.countDown();
    }

    public void setShouldReceiveCoupon(boolean shouldReceive) {
        this.shouldReceiveCoupon = shouldReceive;
    }

    private void giveCoupon(float amount) {
        money += amount;
        System.out.println(String.format("Philosopher %s received a $%.2f coupon. New balance: $%.2f", name,
                amount, money));
    }

    @Override
    public void run() {
        while (money > 0) {
            attemptToDine();
            think();
        }
        System.out.println(String.format("Philosopher %s has left the restaurant for good.", name));
    }

    private void attemptToDine() {
        seat = Optional.ofNullable(Seat.availableSeat());
        if (seat.isPresent() && seat.get().attemptToOccupy()) {
            System.out.println(
                    String.format("Philosopher %s is being seated in chair %d.", name, seat.get().number()));
            think();
            System.out.println(String.format("Philosopher %s is about to call for a waiter.", name));
            boolean hasWaiter = waitForWaiter();
            if (!hasWaiter) {
                System.out.println(
                        String.format("Philosopher %s has left the restaurant without being served.", name));
                seat.get().vacate();
                return;
            }
            System.out.println(String.format("Philosopher %s got waiter, waiting for food.", name));
            shouldReceiveCoupon = false;
            foodServedLatch = new CountDownLatch(1);
            try {
                foodServedLatch.await();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            if (shouldReceiveCoupon) {
                giveCoupon(COUPON_VALUE);
                return; // Leave without eating
            }
            System.out.println(String.format("Philosopher %s got food, about to eat.", name));
            eat();
            pay();
            clearOrder();
        } else {
            System.out.println(String.format("Philosopher %s could not get a seat.", name));
        }
    }

    public void vacateSeat() {
        if (seat.isPresent()) {
            seat.get().vacate();
            seat = Optional.empty(); // Clear the seat reference so philosopher can get a new seat
        }
    }
}
